test/counter_server_example/test/counter_server_example.erl

```diff
@@ -0,0 +1,361 @@
+%%% @doc
+%%%
+%%% Example for using vnet to simulate a cluster of nodes, one of
+%%% which serves as a counter server.
+%%%
+%%% All _scenario functions demonstrate scenarios of using the
+%%% server. See their documentation for more info.
+%%%
+%%% All _no_vnet_scenario functions do not use vnet and their purpose
+%%% is to check the core implementation of the counter server.
+%%%
+%%% @end
+%%%-------------------------------------------------------------------
+
+-module(counter_server_example).
```

Comment:
I think the code in this module should be refactored a bit for readability.

My understanding is that the scenarios are actually permutations of 3 parameters:
1. How to call the server?
    * `no_vnet`: do not use `vnet` at all
    * `rpc`: use `vnet:rpc` calls
    * `proxy`: use `gen_server:call` with `{via, vnet, {server, ?SERVER}}`
2. What should the valid client do?
    * `single`: perform a single request (sometimes called `once` in the code)
    * `double`: perform two requests (sometimes called `twice` in the code)
    * `double_same_gen`: perform two requests assuming the generation is not incremented
3. What complication happens in the test concurrently?
    * no complications
    * `invalid`: an other client makes an invalid request
    * `disconnect`: the connection between the client and server is interrupted
    * `node_down`: the server node is killed

I would recommend consistently naming the scenarios according to the parameters they choose and using generic client/call/check etc. functions that cover all the possible cases and take these parameters as arguments.

Reply:
Sure, leave them as they are.

Maybe even after the cleanup you could leave the original scenarios in the code like this:

```erlang
%% This XY paper refers to this scenario on its legacy name.
%% Please keep this wrapper for the reference of readers of the paper.
old_scenario_fashioned() ->
  new_fashioned_scenario().
```